drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,1026 @@
+#endif
+	ret = H264EncSetCodingCtrl(data->encoder, &codingctrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncSetCodingCtrl error=%d\n", ret);
+		return -1;
+	}
+
+	/* set bit rate configuration */
+	ret = H264EncGetRateCtrl(data->encoder, &ratectrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncGetRateCtrl error=%d\n", ret);
+		return -1;
+	}
+#ifdef VENC_RATECTRL_VBR
+	/* Variable bitrate */
+	ratectrl_cfg.pictureRc = 1;
+	ratectrl_cfg.mbRc = 1;
+	ratectrl_cfg.pictureSkip = 0;
+	ratectrl_cfg.hrd = 0;
+	ratectrl_cfg.qpHdr = VENC_RATECTRL_QP;
+	ratectrl_cfg.qpMin = 10;
+	ratectrl_cfg.qpMax = 51;
+	ratectrl_cfg.gopLen = VENC_RATECTRL_FPS;
+	ratectrl_cfg.bitPerSecond = ((data->width * data->height * 12) * VENC_RATECTRL_FPS) / 30;
+	ratectrl_cfg.intraQpDelta = 0;
+#else
+	/* Constant bitrate */
+	ratectrl_cfg.pictureRc = 0;
+	ratectrl_cfg.mbRc = 0;
+	ratectrl_cfg.pictureSkip = 0;
+	ratectrl_cfg.hrd = 0;
+	ratectrl_cfg.qpHdr = VENC_RATECTRL_QP;
+	ratectrl_cfg.qpMin = VENC_RATECTRL_QP;
+	ratectrl_cfg.qpMax = VENC_RATECTRL_QP;
+#endif
+	ret = H264EncSetRateCtrl(data->encoder, &ratectrl_cfg);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncSetRateCtrl error=%d\n", ret);
+		return -1;
+	}
+
+	LOG_DBG("%s<\n", __func__);
+	return 0;
+}
+
+static int encoder_start(struct video_stm32_venc_data *data)
+{
+	H264EncRet ret;
+	struct video_buffer *output;
+	H264EncIn encIn = {0};
+	H264EncOut encOut = {0};
+
+	LOG_DBG("%s>\n", __func__);
+
+	output = k_fifo_get(&data->fifo_output_in, K_FOREVER);
+
+	encIn.pOutBuf = (u32 *)output->buffer;
+	encIn.busOutBuf = (uint32_t) encIn.pOutBuf;
+	encIn.outBufSize = ALIGNED_SIZE(output->size);
+
+	/* create stream */
+	ret = H264EncStrmStart(data->encoder, &encIn, &encOut);
+	if (ret != H264ENC_OK) {
+		LOG_ERR("H264EncStrmStart error=%d\n", ret);
+		return -1;
+	}
+
+	output->bytesused = encOut.streamSize;
+	LOG_DBG("SPS/PPS generated, size= %d\n", output->bytesused);
+
+	LOG_DBG("%s> %p\n", __func__, &data->fifo_output_out);
+	k_fifo_put(&data->fifo_output_in, output);
+	k_fifo_put(&data->fifo_output_out, output);
+
+	data->resync = true;
+
+	LOG_DBG("%s<\n", __func__);
+	return 0;
+}
+
+static void  encoder_reset(void)
+{
+	/* VENC HW REset */
+	__HAL_RCC_VENC_FORCE_RESET();
+	k_sleep(K_MSEC(1));
+	__HAL_RCC_VENC_RELEASE_RESET();
+	k_sleep(K_MSEC(1));
+}
+
+static int encode_frame(struct video_stm32_venc_data *data)
+{
+	int ret = H264ENC_FRAME_READY;
+	struct video_buffer *input;
+	struct video_buffer *output;
+	H264EncIn encIn = {0};
+	H264EncOut encOut = {0};
+
+	LOG_DBG("%s>\n", __func__);
+
+	if (!data->encoder) {
+		ret = encoder_prepare(data);
+		if (ret)
+			return -1;
+
+		ret = encoder_start(data);
+		if (ret)
+			return -1;
+
+		LOG_DBG("SPS/PPS generated and pushed\n");
+		return 0;
+	}
+
+	input = k_fifo_get(&data->fifo_input, K_NO_WAIT);
+	if (!input)
+		return 0;
+	LOG_DBG("input=%p\n", input);
+
+	output = k_fifo_get(&data->fifo_output_in, K_NO_WAIT);
+	if (!output) {
+		return 0;
```

Comment:
At this line, the `input` buffer is dequeued and dropped by the driver.

There would need some scheme to synchronize both queue to make sure to have both an input and an output buffer poped from the FIFO at the same time.

- One method is to turn it into a thread with `K_FOREVER` instead of `K_NO_WAIT`, and therefre, never `return 0` as they are guaranteed to return something with `K_FOREVER`.

- One method is to 'peek' at the FIFO without popping the buffer. If there is a buffer ready on both FIFO, then go ahead and pop them on both in confidence that none of them would be NULL (as it was checked beforehand)...

Maybe this does not cause an issue in the way the `tcpserversink` uses this driver though!